File: apps/ios/ValueLens/Features/Company/Layouts/ClassicLayout.swift
import SwiftUI

/// What a layout is allowed to see.
///
/// Deliberately narrow: the report, the selected model's result, the core's plain-language summary,
/// and a few callbacks for things only the container can do (present a sheet, switch modes). A
/// layout cannot reach the repository, the network or the core, which is the mechanical reason it
/// cannot compute a number of its own (CLAUDE.md non-negotiable 7).
struct LayoutContext {
    let report: ValuationReport
    let result: ModelResult
    let explain: ExplainSummary?
    /// Expert Mode, or "show me the math" tapped for this company. One shared presentation across
    /// layouts — see `ExpertBody`.
    let expert: Bool
    /// True when Expert Mode is off and the reader turned the math on just for this screen.
    let showingMathTemporarily: Bool
    let model: Binding<ValuationModel>
    let editPrice: () -> Void
    let setShowMath: (Bool) -> Void
    let expandAll: Binding<Bool?>
    let expandVersion: Binding<Int>
    let reportIssue: () -> Void
    /// The core's grades, rules and chips for the active lens. Only the report card reads it; nil
    /// until the core has answered, and a layout must render sensibly without it.
    var reportCard: ReportCardSummary? = nil
    /// Flips Value ⇄ Growth. The lens is shown on the screen it affects, so it is switched there too.
    var swapLens: () -> Void = {}

    var demandedComponents: Set<RequiredComponent> {
        RequiredComponent.demanded(by: report, result: result)
    }
}

/// The layout that shipped in Sprints 0–4, unchanged in substance and in order.
///
/// Two deliberate differences from the pre-Sprint-5 file, both flagged in the commit:
///  1. The inline "value withheld", sector-mode and warnings markup became the shared
///     `RequiredComponents`, so the second layout cannot quietly tell a different story.
///  2. Data notes are no longer expert-only. They were hidden in basic mode, which is exactly the
///     omission `RequiredComponent.dataNotes` exists to prevent.
struct ClassicLayout: View {
    let ctx: LayoutContext

    private var report: ValuationReport { ctx.report }
    private var result: ModelResult { ctx.result }

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            header

            ModelToggle(model: ctx.model,
                        expert: ctx.expert,
                        blurbOverride: ctx.expert ? nil : (ctx.model.wrappedValue == .traditional ? ctx.explain?.blurbA : ctx.explain?.blurbB))

            if let sector = report.sector, sector.mode != "general" {
                SectorModeBadge(sector: sector)
            }

            if report.checksFailed {
                WithheldValueCard(report: report, result: result)
            } else {
                Card {
                    VStack(alignment: .leading, spacing: 12) {
                        if !ctx.expert, let explain = ctx.explain {
                            Text(ctx.model.wrappedValue == .traditional ? explain.verdictA : explain.verdictB)
                                .font(.vlBody)
                                .foregroundStyle(Theme.textPrimary)
                                .fixedSize(horizontal: false, vertical: true)
                        }
                        MarginOfSafetyView(mos: result.marginOfSafety, compact: !ctx.expert)
                    }
                }
                if result.marginOfSafety.verdict == .insufficientData {
                    WithheldValueCard(report: report, result: result)
                }
            }

            DataChecksCard(checks: report.dataChecks,
                           summary: ctx.explain?.checksSummary ?? "\(report.dataChecks.count) checks run",
                           expanded: ctx.expert)

            if let coverage = report.coverage, !coverage.gaps.isEmpty {
                CoverageGapCard(coverage: coverage, expert: ctx.expert, onReport: ctx.reportIssue)
            }

            // Was expert-only before Sprint 5; a basic-mode reader was not told about MCD's
            // share-scale correction or BRK-B's derived EPS at all.
            DataNotesSection(report: report, includeChecks: false)

            if ctx.expert {
                ExpertBody(ctx: ctx)
            } else {
                basicBody
            }

            ProvenanceRow(report: report)

            Text(report.disclaimer)
                .font(.caption2)
                .foregroundStyle(Theme.textTertiary)
                .multilineTextAlignment(.center)
                .frame(maxWidth: .infinity)
        }
        .padding(16)
    }

    private var header: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(report.company.name).font(.vlTitle).foregroundStyle(Theme.textPrimary)
            HStack(alignment: .firstTextBaseline, spacing: 12) {
                Button(action: ctx.editPrice) {
                    HStack(spacing: 4) {
                        Text(Fmt.money(report.price)).font(.vlDisplay).foregroundStyle(Theme.price)
                        Image(systemName: "pencil.circle").font(.caption).foregroundStyle(Theme.textTertiary)
                    }
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Market price \(Fmt.money(report.price)). Tap to override.")
                VStack(alignment: .leading, spacing: 0) {
                    Text(report.quote.map { "\($0.source) · \(Fmt.shortDate($0.asOf))" } ?? "no quote — tap to enter")
                        .font(.caption2).foregroundStyle(Theme.textTertiary)
                    if ctx.expert {
                        Text("CIK " + String(report.company.cik)).font(.caption2).foregroundStyle(Theme.textTertiary)
                    }
                }
            }
        }
    }

    @ViewBuilder
    private var basicBody: some View {
        SectionHeader(title: "How healthy is the business?", subtitle: "Tap a tile for a plain-English explanation")
        LazyVGrid(columns: [GridItem(.flexible()), GridItem(.flexible())], alignment: .leading, spacing: 8) {
            ForEach(ctx.explain?.facts ?? []) { FactTile(fact: $0) }
        }
        Button { withAnimation { ctx.setShowMath(true) } } label: {
            Text("Show me the math").frame(maxWidth: .infinity)
        }
        .buttonStyle(.bordered)
        .tint(Theme.accent)
        Text("Turn on Expert Mode in Settings to always see formulas and SEC line items.")
            .font(.caption)
            .foregroundStyle(Theme.textTertiary)
            .multilineTextAlignment(.center)
            .frame(maxWidth: .infinity)
    }
}

File: apps/android/app/src/main/java/com/swcsoftware/valuelens/ui/layouts/ClassicLayout.kt
package com.swcsoftware.valuelens.ui.layouts

import android.content.Context
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.swcsoftware.valuelens.AppState
import com.swcsoftware.valuelens.domain.CompanyRef
import com.swcsoftware.valuelens.domain.EngineException
import com.swcsoftware.valuelens.domain.HistoryPoint
import com.swcsoftware.valuelens.domain.ModelResult
import com.swcsoftware.valuelens.domain.ValuationModel
import com.swcsoftware.valuelens.domain.ValuationReport
import com.swcsoftware.valuelens.domain.Verdict
import com.swcsoftware.valuelens.domain.price
import com.swcsoftware.valuelens.domain.result
import com.swcsoftware.valuelens.domain.verdictEnum
import com.swcsoftware.valuelens.core.Coverage
import com.swcsoftware.valuelens.core.Explain
import com.swcsoftware.valuelens.export.Exporter
import com.swcsoftware.valuelens.ui.components.CoverageGapCard
import com.swcsoftware.valuelens.ui.components.DataChecksCard
import com.swcsoftware.valuelens.ui.components.FactTile
import com.swcsoftware.valuelens.ui.Fmt
import com.swcsoftware.valuelens.ui.components.Card
import com.swcsoftware.valuelens.ui.components.MarginOfSafetyView
import com.swcsoftware.valuelens.ui.components.MetricRow
import com.swcsoftware.valuelens.ui.components.ModelToggle
import com.swcsoftware.valuelens.ui.components.SectionHeader
import com.swcsoftware.valuelens.ui.components.ThinDivider
import com.swcsoftware.valuelens.ui.theme.VL
import kotlinx.coroutines.launch
import com.swcsoftware.valuelens.domain.CoverageReport
import com.swcsoftware.valuelens.ui.screens.SnapshotGrid
import com.swcsoftware.valuelens.ui.screens.HistoryChart
import com.swcsoftware.valuelens.ui.screens.HistoryTable
import com.swcsoftware.valuelens.core.ReportCardSummary
import com.swcsoftware.valuelens.ui.LayoutStyle
import com.swcsoftware.valuelens.ui.RequiredComponent

/**
 * What a layout is allowed to see.
 *
 * Deliberately narrow: the report, the selected model's result, and a few callbacks for the things
 * only the screen can do. A layout cannot reach the repository, the network or the core, which is
 * the mechanical reason it cannot compute a number of its own (CLAUDE.md non-negotiable 7).
 *
 * Mirrors apps/ios/ValueLens/Features/Company/Layouts/ClassicLayout.swift.
 */
data class LayoutContext(
    val report: ValuationReport,
    val result: ModelResult,
    /** Expert Mode, or "show me the math" tapped for this company. */
    val expert: Boolean,
    /** True when Expert Mode is on globally, so "hide the math" should not be offered. */
    val expertModeAlwaysOn: Boolean,
    val model: ValuationModel,
    val onModel: (ValuationModel) -> Unit,
    val onEditPrice: () -> Unit,
    val onSetShowMath: (Boolean) -> Unit,
    val onReportIssue: (CoverageReport) -> Unit,
    /** The core's grades, rules and chips for the active lens. Only the report card reads it. */
    val reportCard: ReportCardSummary? = null,
    /** Flips Value ⇄ Growth; the lens is shown on the screen it affects, so it is switched there. */
    val onSwapLens: () -> Unit = {},
) {
    val demandedComponents: Set<RequiredComponent>
        get() = RequiredComponent.demandedBy(report, result)
}

/** The layout that shipped in Sprints 0–4, unchanged in substance and in order. */
@Composable
fun ClassicLayout(ctx: LayoutContext) {
    val r = ctx.report
    val res = ctx.result
    val expert = ctx.expert
    val model = ctx.model

    Column(Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(16.dp, 8.dp, 16.dp, 32.dp)) {
val failed = r.dataChecks.any { it.status == "fail" }
Text(r.company.name, style = MaterialTheme.typography.headlineSmall, color = VL.textPrimary)
Row(verticalAlignment = Alignment.Bottom) {
    Text(Fmt.money(r.price), fontSize = 38.sp, fontWeight = FontWeight.ExtraBold, color = VL.price, modifier = Modifier.clickable { ctx.onEditPrice() })
    Column(Modifier.padding(start = 12.dp, bottom = 8.dp)) {
        Text(r.quote?.let { "${it.source} · ${it.asOf.take(10)}" } ?: "no quote — tap to enter", fontSize = 11.sp, color = VL.textTertiary)
        if (expert) Text("CIK ${r.company.cik}", fontSize = 11.sp, color = VL.textTertiary)
    }
}
Spacer(Modifier.height(10.dp))
ModelToggle(model, labels = ValuationModel.entries.map { Explain.modelName(it == ValuationModel.A, expert) }, onChange = ctx.onModel)
if (!expert) Text(Explain.modelBlurb(model == ValuationModel.A, r.sector?.mode ?: "general"), style = MaterialTheme.typography.bodySmall, color = VL.textSecondary, modifier = Modifier.padding(vertical = 6.dp))
r.sector?.takeIf { it.mode != "general" }?.let { SectorModeBadge(it) }
Spacer(Modifier.height(8.dp))

if (failed) {
            WithheldValueCard(r, res)
        } else {
    Card { if (!expert) Text(Explain.verdictSentence(r, res), color = VL.textPrimary, modifier = Modifier.padding(bottom = 12.dp)); MarginOfSafetyView(res.marginOfSafety, compact = !expert) }
    if (res.marginOfSafety.verdictEnum == Verdict.INSUFFICIENT) { Spacer(Modifier.height(12.dp)); WithheldValueCard(r, res) }
}
Spacer(Modifier.height(12.dp))
DataChecksCard(r.dataChecks, Explain.checksSummary(r), expert)
r.coverage?.takeIf { it.gaps.isNotEmpty() }?.let { cov ->
    Spacer(Modifier.height(12.dp))
    CoverageGapCard(cov, expert) { ctx.onReportIssue(cov) }
}

if (!expert) {
    SectionHeader("How healthy is the business?", "Tap a tile for a plain-English explanation")
    val facts = Explain.healthFacts(r)
    facts.chunked(2).forEach { row -> Row(Modifier.fillMaxWidth().padding(top = 8.dp), horizontalArrangement = Arrangement.spacedBy(8.dp)) { row.forEach { f -> FactTile(f.label, f.value, f.tone, f.plain, Modifier.weight(1f)) }; if (row.size == 1) Spacer(Modifier.weight(1f)) } }
    Spacer(Modifier.height(16.dp))
    TextButton({ ctx.onSetShowMath(true) }, Modifier.fillMaxWidth()) { Text("Show me the math", color = VL.accent) }
    Text("Turn on Expert Mode in Settings to always see formulas and SEC line items.", style = MaterialTheme.typography.bodySmall, color = VL.textTertiary, modifier = Modifier.fillMaxWidth(), textAlign = androidx.compose.ui.text.style.TextAlign.Center)
} else {
    ExpertBody(ctx)
}
Text(r.disclaimer, style = MaterialTheme.typography.bodySmall, color = VL.textTertiary, modifier = Modifier.fillMaxWidth().padding(top = 20.dp), textAlign = androidx.compose.ui.text.style.TextAlign.Center)
    }
}
